src/components/views/dialogs/InviteDialog.tsx

```diff
@@ -1215,8 +1215,11 @@ export default class InviteDialog extends React.PureComponent<Props, IInviteDial
         }
     };
 
-    private onTabChange = (tabId: TabId): void => {
-        this.setState({ currentTabId: tabId });
+    private onTabChange = (tabId: string): void => {
+        // Tabs that this function can be called with are generated in the `render()` method of this
+        // class. When generating each tab, the `TabId` enum is used to generate the corresponding
+        // tabId so the parameter will always be a string from the `TabId` enum.
```

Comment:
So you can't just replace `string` with `TabId` in a few places?

Reply:
Unfortunately not at least according to my repeated attempts. I end up going down a slippery slope of making both the `Tab` and `TabbedView` classes generic, which adds a lot of code and clutter for very little benefit (no other users of this class use anything other than strings, from what I can see).

I'll have one more pop tomorrow morning and see if there's some neat solution I'm missing

Reply:
Tried again this morning and couldn't find a small solution to this. Will leave the explanatory comment in to cover why we're using `as` there

Reply:
I've got a diff to make it generic, its not too bad.

```patch
Index: src/components/structures/TabbedView.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/components/structures/TabbedView.tsx b/src/components/structures/TabbedView.tsx
--- a/src/components/structures/TabbedView.tsx	(revision 2dc171e392ba5ee15e23ab6c8179390e2904bb28)
+++ b/src/components/structures/TabbedView.tsx	(date 1682586369353)
@@ -29,7 +29,7 @@
 /**
  * Represents a tab for the TabbedView.
  */
-export class Tab {
+export class Tab<T extends string> {
     /**
      * Creates a new tab.
      * @param {string} id The tab's ID.
@@ -39,7 +39,7 @@
      * @param {string} screenName The screen name to report to Posthog.
      */
     public constructor(
-        public readonly id: string,
+        public readonly id: T,
         public readonly label: string,
         public readonly icon: string | null,
         public readonly body: React.ReactNode,
@@ -52,20 +52,20 @@
     TOP = "top",
 }
 
-interface IProps {
-    tabs: NonEmptyArray<Tab>;
-    initialTabId?: string;
+interface IProps<T extends string> {
+    tabs: NonEmptyArray<Tab<T>>;
+    initialTabId?: T;
     tabLocation: TabLocation;
-    onChange?: (tabId: string) => void;
+    onChange?: (tabId: T) => void;
     screenName?: ScreenName;
 }
 
-interface IState {
-    activeTabId: string;
+interface IState<T extends string> {
+    activeTabId: T;
 }
 
-export default class TabbedView extends React.Component<IProps, IState> {
-    public constructor(props: IProps) {
+export default class TabbedView<T extends string> extends React.Component<IProps<T>, IState<T>> {
+    public constructor(props: IProps<T>) {
         super(props);
 
         const initialTabIdIsValid = props.tabs.find((tab) => tab.id === props.initialTabId);
@@ -78,7 +78,7 @@
         tabLocation: TabLocation.LEFT,
     };
 
-    private getTabById(id: string): Tab | undefined {
+    private getTabById(id: T): Tab<T> | undefined {
         return this.props.tabs.find((tab) => tab.id === id);
     }
 
@@ -87,7 +87,7 @@
      * @param {Tab} tab the tab to show
      * @private
      */
-    private setActiveTab(tab: Tab): void {
+    private setActiveTab(tab: Tab<T>): void {
         // make sure this tab is still in available tabs
         if (!!this.getTabById(tab.id)) {
             if (this.props.onChange) this.props.onChange(tab.id);
@@ -97,7 +97,7 @@
         }
     }
 
-    private renderTabLabel(tab: Tab): JSX.Element {
+    private renderTabLabel(tab: Tab<T>): JSX.Element {
         const isActive = this.state.activeTabId === tab.id;
         const classes = classNames("mx_TabbedView_tabLabel", {
             mx_TabbedView_tabLabel_active: isActive,
@@ -130,11 +130,11 @@
         );
     }
 
-    private getTabId(tab: Tab): string {
+    private getTabId(tab: Tab<T>): string {
         return `mx_tabpanel_${tab.id}`;
     }
 
-    private renderTabPanel(tab: Tab): React.ReactNode {
+    private renderTabPanel(tab: Tab<T>): React.ReactNode {
         const id = this.getTabId(tab);
         return (
             <div className="mx_TabbedView_tabPanel" key={id} id={id} aria-labelledby={`${id}_label`}>
```

and then just needs a handful of tiny updates wherever `Tab` is used externally

Reply:
I'm pleased to see this is what I had in my first attempt - in which case I'll put this in

Reply:
@alunturner handled in https://github.com/matrix-org/matrix-react-sdk/pull/10726

Reply:
Though I'm more than happy for you to land it in this PR instead, up to you!

Reply:
Ahhh I see - I misunderstood and thought you were helping me out with some code. I'll merge it in as part of this so long as that causes you no issues.

The changes you've made in that PR make the effort of making the classes generic more worthwhile. Only tiny difference I've added is a default of `<T extends string = string>` on the `Tab` class to prevent people having to specify that case.

Reply:
I worry that default will yield too soft types, letting you pass tabs which have TabId `string` rather than a specific enum/union type

Reply:
Ok cool - in that case it'll make more sense for me to roll back to one of my earlier commits and allow #10726 to handle this.